### ISSUE TITLE
Add built-in Playnite overlay

### ISSUE THREAD

**Issue by PopeRigby** [JosefNemec/Playnite#1390]
To have an overlay in a non-Steam (or another launcher with an overlay) game, you have to add every single game as a shortcut to get the overlay. This makes it frustrating for people who want an overlay in every game, but don't want to fiddle with another launcher's overlay. So a built-in Playnite overlay with the time, a web browser and #1264 integrated would be super useful.

**Comment by PopeRigby** [comment]
I think using Firefox (or a version of it) would be a good idea for the browser. Gecko is very fast.

**Comment by kvdrrrrr** [comment]
@PopeRigby Playnite already uses CefSharp, why introduce GeckoFX as second dependency for the same tasks? Also, CefSharp claims to [have superior performance](https://cefsharp.github.io/).

**Comment by PopeRigby** [comment]
Why help the Chromium monopoly?

**Comment by JosefNemec** [comment]
CefSharp is by far more mature project with superior WPF integration (at least it was when I was integrating it into Playnite), which is critical for Playnite.

Overall don't expect in-game overlay any time soon since it's quite complex feature and I don't have necessary OpenGL, DirectX and Vulkan knowledge to implement it reliably and efficiently (and currently I don't have enough time learn it, although I would like to). Somebody else would have to do it.

**Comment by kvdrrrrr** [comment]
@PopeRigby That's opinionated. I'm not necessarily saying that CefSharp is faster than GeckoFX. It would just take a tremendous amount of work to use it, because Playnite already uses CefSharp. If anything, this should be made a seperate issue, since it's not much related with in-game overlays.

**Comment by PopeRigby** [comment]
I'll admit I don't know much about web rendering engines. I was suggesting Gecko because it seems to be very fast and Chromium has a monopoly on the engine market.

**Comment by trymeouteh** [comment]
If you do go down the Chromium path, please use ungoogled-chromium since it removes the tracking and all the google spyware

**Comment by darklinkpower** [comment]
The best way to go for this in my opinion would be to implement a way to use the steam overlay for all games in a similar approach to this https://forums.launchbox-app.com/topic/43142-steamlauncher/

Mantaining, bug fixing and developing for all the different Graphics APIs would be a nightmare, why not just use something that already exists and is developed by another party and has all the features people want? An exclusive Overlay, while a good idea, would be a waste of resources considering Crow is almost the sole developer of Playnite.

**Comment by JosefNemec** [comment]
Launching all games via Steam would come with some serious disadvantages (main in terms of play time tracking). I also don't like an idea of some Playnite feature being directly tied to another launcher (unless it's a game import).

**Comment by PopeRigby** [comment]
> I also don't like an idea of some Playnite feature being directly tied to another launcher (unless it's a game import).
Yeah. Also, Steam's overlay isn't that good. An open source overlay would be better.

**Comment by darklinkpower** [comment]
Yeah I agree, an exclusive overlay would be the best option. I was just thinking that developing a new one from scratch would be fairly difficult, maybe you would need a dedicated developer focused exclusively at least for a while in the overlay and with enough knowledge of Graphics API to make one and mantaining it. I mean, games can run in vulkan, the different DirectX versions, OpenGL and that means a whole lot of issues. I don't know anything about coding but that's just my guess.

**Comment by kvdrrrrr** [comment]
@darklinkpower It might be worth to take a look at existing projects which already utilize multiple Graphics API hooking.

[ImGUI contains a bunch of examples, excluding the actual hooking](https://github.com/ocornut/imgui/tree/master/examples)

[ReShade's code contains pretty much everything what's needed](https://github.com/crosire/reshade/tree/master/source)

I'm not exactly sure how would hooking work in a managed process. Would a library such as EasyHook be enough to accomplish this or it would be needed to implement a native helper binary?

**Comment by PopeRigby** [comment]
Mumble has an open source overlay. I haven't ever taken a look at it, but could it be used for reference?

**Comment by gagrecco** [comment]
It would be cool to have the ability to close the game when you launch the overlay. Also have the ability to load/save state

**Comment by LeonPro12** [comment]
Is there any progress?

**Comment by JosefNemec** [comment]
> Is there any progress?

No, my original comment is still relevant https://github.com/JosefNemec/Playnite/issues/1390#issuecomment-528863373

**Comment by jmvallejo** [comment]
I'd love to have an overlay, mainly because I want to use playnite for emulators and having an overlay for closing the emulator / terminating the process would be awesome in fullscreen mode.

**Comment by ZazaMastro** [comment]
> Update on the overlay situation. Since there are no devs available with enough expertise to implement this reliably, I decided to implement this as Xbox Game Bar widget. This gives us reliable way of injecting overlay into games (not having to bother with custom injection, which can be problematic, especially in games using anti-cheat system) and any Playnite related functionality (for example to close running emulator, which seems to be the most requested use for an overlay) can be relatively easily added via custom widget. Still no ETA on when the widget will be ready.

This is a fantastic news. At least we will get the needed functionalities directly from gamepad.

**Comment by Tyestor** [comment]
@poperigby don't worry, josef knows even less than you "cefsharp is more mature" lmao gecko has been around since 1997 mate and anyone with a brain should trust mozilla over alphabet.

"Launching all games via Steam would come with some serious disadvantages" yeah no shit, that would defeat the purpose of playnite. but most people are doing that anyway, so why use playnite again? lmao. at least you know that much though. what are you doing about the fact you need 2-3 extra programs to use playnite properly? ds4windows/rewasd/special k/controller companion/joytokey/joyxkey/dualsensex/xpadder/hidhide/x360ce/GlosSI

**Comment by JosefNemec** [comment]
@Tyestor 
1) I was not talking about CEF vs Gecko, I was talking about CefSharp and GeckoFX, which are wrappers for those frameworks to be usable together in .NET via WPF or WinForms. CefSharp is way more usable in that sense, that's just a fact, no discussion about it.

2) Most people by what metric? Even I don't have any numbers for that. Obviously, long term plan is to have more things integrated for ease of use (like DirectInput support), but this a project still maintained primarily by a single person and it's already quite big to maintain. So if there's usable 3rd party solution, I will not rush to create different implementation for Playnite.

3) Please stay on topic related to this issue. I would also try less condescending tone in future if you want your feedback to be actually taken seriously.

**Comment by aliceroherty** [comment]
Just started using playnite in 2025 and would love to see this feature added. I loved using Steam's big picture mode but found playnite to be a much better alternative in all areas except just this one. Would be so nice to have a menu come up when pressing the ps/xbox button to close the game or return to the library etc. 

Debated trying to contribute to this feature but it seems PR's are on hold while a full rewrite takes place.

Not sure if there is an ETA on the rewrite or if maybe this feature will be included as part of it but feel free to mention me if/when adding this feature becomes a possibility I'd be willing to take a crack at it.

**Comment by JosefNemec** [comment]
Overlay won't be part of P11 release but it's something I plan to look into some time after. There's also somebody on Discord working on an overlay as a plugin for current Playnite release.

**Comment by aliceroherty** [comment]
> Overlay won't be part of P11 release but it's something I plan to look into some time after. There's also somebody on Discord working on an overlay as a plugin for current Playnite release.

Oh that's good to know, do you know where I could find the repo for that plug-in? Maybe they could use a hand over there.

**Comment by JosefNemec** [comment]
https://github.com/jonosellier/PlayniteGameOverlay
